### ISSUE TITLE
Code action to add all missing imports

### ISSUE THREAD

**Issue by incendial** [Dart-Code/Dart-Code#4458]
**Is your feature request related to a problem? Please describe.**
Importing multiple entities one by takes a lot more time that I'd expect.

**Describe the solution you'd like**
Add a code action that (if multiple entities can be imported) would allow to import all of them at once.

Similar feature is available for TS and works amazing:
<img width="375" alt="Screenshot 2023-03-24 at 22 29 21" src="https://user-images.githubusercontent.com/21116742/227610041-f62e02b1-d94f-4a36-b2c4-363abc963b70.png">

**Comment by DanTup** [comment]
@bwilkerson @pq any thoughts on this? It's possible that some diagnostics have multiple possible fixes:

<img width="488" alt="Screenshot 2023-03-28 at 10 34 50" src="https://user-images.githubusercontent.com/1078012/228194480-611f7b10-69f1-4dff-b26b-df02aa3801bd.png">

TypeScript just always uses the first one, but I'm not convinced that's the best decision. I think fixing only those where there's a single option might be a reasonable (or even better) choice.

I'm not sure whether the "fix in file" functionality can handle cases like this, but I agree it would be useful to have (particularly for copy/pasting code while VS Code still hasn't finalized APIs for copy/paste that would let use handle the imports automatically).

**Comment by incendial** [comment]
@DanTup also have a slightly related question: I have several lib entry point export files, is there an option to exclude them from the import suggestions? It's really easy to create a cyclic imports chain in the project and it frustrates me a lot 😞. Especially when they are chosen as the first option or appear first in the code completion suggestions.

**Comment by DanTup** [comment]
@incendial can you give a concrete example? (perhaps in its own issue here or in the SDK).

**Comment by incendial** [comment]
@DanTup sure, done https://github.com/Dart-Code/Dart-Code/issues/4468

**Comment by bwilkerson** [comment]
> It's possible that some diagnostics have multiple possible fixes ...

Yes. That's not uncommon.

> TypeScript just always uses the first one, but I'm not convinced that's the best decision.

It's almost certainly not the best choice for the analysis server because we haven't done any work to optimize the order in which the fixes appear.

> I'm not sure whether the "fix in file" functionality can handle cases like this ...

Not as written today. Each correction producer can say whether it's enabled for fix-in-file mode, but that would just add the fixes produced by the producer for each reported diagnostic. There's no logic to support choosing one from potentially multiple fixes. If we can define an algorithm for making that choice then we could implement the rest of the required functionality, but today we don't have that algorithm.

> ... while VS Code still hasn't finalized APIs for copy/paste that would let use handle the imports automatically

I didn't realize that that functionality was even being discussed. That's one of the things I miss most in our tooling, so it's fantastic if they're actually thinking about it.

**Comment by DanTup** [comment]
> There's no logic to support choosing one from potentially multiple fixes. If we can define an algorithm for making that choice then we could implement the rest of the required functionality, but today we don't have that algorithm.

For this particular case, I thought maybe it'd be fine to _only_ work if there's only a _single_ fix. Although it occurs to me that a single import might have both a relative and `package:` version. So it's possible there are (at least) two fixes in all cases. We do sort them so the preferred one is on top (using lints as a hint), but that's a detail the fix processor probably wouldn't understand.


> I didn't realize that that functionality was even being discussed. That's one of the things I miss most in our tooling, so it's fantastic if they're actually thinking about it.

Yeah, the relevant issue is https://github.com/microsoft/vscode/issues/30066 (apparently it was some time ago that I raised it, when you first mentioned adding support to the server 😄). It's assigned to the current milestone, although that's been the case for many months so it's not a good indicator of it coming. We'll need it to get into LSP too before we can use it, but we might be able to contribute towards that once it's in VS Code (although, I think LSP might be close to a release, so the subsequent one might also be a while off).

**Comment by Abdelazeem777** [comment]
Any updates or workarounds?
